src/algorithms/pid/ParticlesWithPID.cc

```diff
@@ -0,0 +1,329 @@
+
+            // calculate dist(eta,phi)
+            auto match_dist = std::hypot(
+                    in_part_eta - in_track_eta,
+                    in_part_phi - in_track_phi
+                    );
+
+            // check if the match is close enough: within user-specified tolerances
+            auto match_is_close =
+                std::abs(in_part_eta - in_track_eta) < m_cfg.etaTolerance &&
+                std::abs(in_part_phi - in_track_phi) < m_cfg.phiTolerance;
+            if (match_is_close)
+                prox_match_list.push_back(ProxMatch{match_dist, in_pid_idx});
+
+            // logging
+            m_log->trace("  - (eta,phi) = ( {:>5.4}, {:>5.4} deg ),  match_dist = {:<5.4}{}",
+                    in_track_eta,
+                    in_track_phi * 180.0 / M_PI,
+                    match_dist,
+                    match_is_close ? " => CLOSE!" : ""
+                    );
+
+        } // end loop over input CherenkovParticleID objects
+
+        // check if at least one match was found
+        if (prox_match_list.size() == 0) {
+            m_log->warn("no matching CherenkovParticleID found for this particle");
+            return false;
+        }
+
+        // choose the closest matching CherenkovParticleID object corresponding to this input reconstructed particle
+        auto closest_prox_match = *std::min_element(
+                prox_match_list.begin(),
+                prox_match_list.end(),
+                [] (ProxMatch a, ProxMatch b) { return a.match_dist < b.match_dist; }
+                );
+        auto in_pid_matched = in_pids.at(closest_prox_match.pid_idx);
+        m_log->trace("  => best match: match_dist = {:<5.4} at idx = {}",
+                closest_prox_match.match_dist,
+                closest_prox_match.pid_idx
+                );
+
+        // convert `CherenkovParticleID` object's hypotheses => set of `ParticleID` objects
+        auto out_pid_index_map = ConvertParticleID::ConvertToParticleIDs(in_pid_matched, out_pids, true);
+        if (out_pid_index_map.size() == 0) {
+            m_log->error("found CherenkovParticleID object with no hypotheses");
+            return false;
+        }
+
+        // relate matched ParticleID objects to output particle
+        bool first = true;
+        for (const auto& [out_pids_index, out_pids_id] : out_pid_index_map) {
+            const auto& out_pid = out_pids->at(out_pids_index);
+            if (out_pid.id() != out_pids_id) { // sanity check
+                m_log->error("indexing error in `edm4eic::ParticleID` collection");
+                return false;
+            }
+            in_part.addToParticleIDs(out_pid);
+            if (first) {
+                in_part.setParticleIDUsed(out_pid); // highest likelihood is the first
+                in_part.setGoodnessOfPID(1); // FIXME: not used yet, aside from 0=noPID vs 1=hasPID
+                first = false;
+            }
+        }
```

Comment:
At this point `out_pid_index_map` is non-empty, so can we do this?
```suggestion
        for (const auto& [out_pids_index, out_pids_id] : out_pid_index_map) {
            const auto& out_pid = out_pids->at(out_pids_index);
            if (out_pid.id() != out_pids_id) { // sanity check
                m_log->error("indexing error in `edm4eic::ParticleID` collection");
                return false;
            }
            in_part.addToParticleIDs(out_pid);
        }
        in_part.setParticleIDUsed(in_part.getParticleIDs().front()); // highest likelihood is the first
        in_part.setGoodnessOfPID(1); // FIXME: not used yet, aside from 0=noPID vs 1=hasPID
```
Or is there a future direction that motivates the use of `first`?

Reply:
Your suggestion is better